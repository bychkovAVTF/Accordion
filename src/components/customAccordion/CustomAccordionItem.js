// подключение компонентов из библиотеки MUI
import Accordion from "@mui/material/Accordion";
import AccordionDetails from "@mui/material/AccordionDetails";
import AccordionSummary from "@mui/material/AccordionSummary";
import Typography from "@mui/material/Typography";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";

// деструктуризации объекта входных данных
export const CustomAccordionItem = ({ isExpanded, onActionChange, title, text }) => {
  return (
    // передача информации открыт/закрыт - isExpanded
    // передача действия при нажатии - onActionChange
    <Accordion expanded={isExpanded} onChange={() => onActionChange()}>
      <AccordionSummary expandIcon={<ExpandMoreIcon />}>
        {/* вывод заголовка пункта */}
        <Typography sx={{ width: "33%", flexShrink: 0 }}>{title}</Typography>
      </AccordionSummary>
      <AccordionDetails>
        {/* вывод текста пункта */}
        <Typography>{text}</Typography>
      </AccordionDetails>
    </Accordion>
  );
};
