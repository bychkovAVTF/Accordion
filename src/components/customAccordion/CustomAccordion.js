import { useState } from "react"; // подключение хука состояния
import { CustomAccordionItem } from "./CustomAccordionItem.js"; // подключение компонента пунтка аккордеона

// экспорт компонента аккордеона
// деструктуризация объекта входных данных {data}
export const CustomAccordion = ({ data }) => {
  // инициализация хука состояния
  // expanded значение состояния
  // setExpanded - функция для изменения состояния
  // false - начальное состояние expanded
  const [expanded, setExpanded] = useState(false);

  // функция для открытия/закрытия пунтка аккордеона
  // если номер панели совпадает с expanded - меняем состояние на false
  // иначе expanded присваиваем номер панели
  const handleChange = (panel) => {
    panel === expanded ? setExpanded(false) : setExpanded(panel);
  };

  // если массив данных не был передан(null) или оказался пустым, то не отображаем аккордеон
  if (!data || data.length === 0) return null;

  return (
    // компонет оболочка
    <div className="Accordion">
      {/* через метод массива map возвращаем массив пунктов аккордеона */}
      {data.map((elem, key) => (
        // компонент пункт аккордеона c входными данными
        <CustomAccordionItem
          // ключ итерации
          key={key}
          // булевое значение открыт/закрыт
          isExpanded={expanded === key}
          // событие при нажатии
          onActionChange={() => handleChange(key)}
          // заголовок пункта
          title={elem.title}
          // описание пункта
          text={elem.description}
        />
      ))}
    </div>
  );
};
