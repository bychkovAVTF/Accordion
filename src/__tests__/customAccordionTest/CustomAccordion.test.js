import { render, screen } from "@testing-library/react"; // подклчаем функции тестирования

import { CustomAccordion } from "../../components/customAccordion/CustomAccordion"; // подключение компонента аккордеона
import { AccordionData } from "../../constants/accordion.const"; // подключение массива данных пунктов

const accordionTestData = [AccordionData[0]]; // создаем тестовый массив данных из одного элемента

// название тестов
describe("Accordion component", () => {
  // проверка на отображение компонента с данными
  it("Accordion render", () => {
    // отображение компонента
    render(<CustomAccordion data={accordionTestData} />);
    // проверка содержит ли компонент загловок 
    expect(screen.getByText(accordionTestData[0].title)).toBeInTheDocument();
    // проверка содержит ли компонент описание 
    expect(screen.getByText(accordionTestData[0].description)).toBeInTheDocument();
  });

  // проверка если в компонент были переданы данные или был передан пустой массив
  it("Accordion render without data", () => {
    // отрис компонента
    render(<CustomAccordion />);
    // проверка что заголовок не отображается
    expect(screen.queryByText(accordionTestData[0].title)).toBeNull();
    // проверка что описание не отображается
    expect(screen.queryByText(accordionTestData[0].description)).toBeNull();
  });

  it("Accordion snapshot", () => {
    // отображение компонента и сохранение его в переменную 
    const accordion = render(<CustomAccordion data={accordionTestData} />);
    // создаем снимок тегов компонента
    expect(accordion).toMatchSnapshot();
  });
});
