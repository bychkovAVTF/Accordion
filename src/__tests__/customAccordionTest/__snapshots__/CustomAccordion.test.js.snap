// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component Accordion snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Accordion"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
        >
          <div
            aria-expanded="false"
            class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-1m4ljmh-MuiTypography-root"
              >
                title 1
              </p>
            </div>
            <div
              class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ExpandMoreIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
                />
              </svg>
            </div>
          </div>
          <div
            class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
            style="min-height: 0px;"
          >
            <div
              class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
            >
              <div
                class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
              >
                <div
                  class="MuiAccordion-region"
                  role="region"
                >
                  <div
                    class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                  >
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                    >
                      Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Accordion"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiAccordion-root MuiAccordion-rounded MuiAccordion-gutters css-1elwnq4-MuiPaper-root-MuiAccordion-root"
      >
        <div
          aria-expanded="false"
          class="MuiButtonBase-root MuiAccordionSummary-root MuiAccordionSummary-gutters css-sh22l5-MuiButtonBase-root-MuiAccordionSummary-root"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiAccordionSummary-content MuiAccordionSummary-contentGutters css-o4b71y-MuiAccordionSummary-content"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-1m4ljmh-MuiTypography-root"
            >
              title 1
            </p>
          </div>
          <div
            class="MuiAccordionSummary-expandIconWrapper css-yw020d-MuiAccordionSummary-expandIconWrapper"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
        </div>
        <div
          class="MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner MuiCollapse-vertical css-9l5vo-MuiCollapse-wrapperInner"
            >
              <div
                class="MuiAccordion-region"
                role="region"
              >
                <div
                  class="MuiAccordionDetails-root css-15v22id-MuiAccordionDetails-root"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
                  >
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
